pymc_experimental/model/marginal_model.py

```diff
@@ -244,6 +248,127 @@ def marginalize(self, rvs_to_marginalize: Union[TensorVariable, Sequence[TensorV
         # Raise errors and warnings immediately
         self.clone()._marginalize(user_warnings=True)
 
+    def unmarginalize(
+        self, idata, var_names=None, include_samples=False, extend_inferencedata=True
+    ):
+        """Computes log-likelihoods of marginalized variables conditioned on parameters
+        of the model given InferenceData with posterior group
+
+        Parameters
+        ----------
+        idata : InferenceData
+            InferenceData with posterior group
+        var_names : sequence of str, optional
+            List of Observed variable names for which to compute log_likelihood. Defaults to all observed variables
+        include_samples : bool, default False
+            Include samples of the marginalized variables
+        extend_inferencedata : bool, default True
+            Whether to extend the original InferenceData or return a new one
+
+        Returns
+        -------
+        idata : InferenceData
+            InferenceData with var_names added to posterior
+
+        """
+        if var_names is None:
+            var_names = self.marginalized_rvs
+
+        joint_logp = self.logp()
+        posterior = idata.posterior
+
+        # Remove Deterministics
+        posterior_values = posterior[
+            [rv.name for rv in mm.free_RVs if rv not in self.marginalized_rvs]
+        ]
+
+        sample_dims = ("chain", "draw")
+        posterior_pts, stacked_dims = dataset_to_point_list(posterior_values, sample_dims)
+        rv_dict = {}
+        rv_dims_dict = {}
+
+        for rv in var_names:
+            m = self.clone()
+            m.register_rv(rv, name=rv.name)
+            m.marginalized_rvs = [r for r in m.marginalized_rvs if r is not rv]
+
+            rv_shape = constant_fold(tuple(rv.shape))
+            rv_domain = get_domain_of_finite_discrete_rv(rv)
+            rv_domain_tensor = pt.swapaxes(
+                pt.full(
+                    (*rv_shape, len(rv_domain)),
+                    rv_domain,
+                    dtype=rv.dtype,
+                ),
+                axis1=0,
+                axis2=-1,
+            )
+
+            marginalized_value = m.rvs_to_values[rv]
+
+            other_values = [v for v in m.value_vars if v is not marginalized_value]
+
+            # TODO: Handle constants
+            # TODO: Handle transformed variables
+            joint_logp_op = OpFromGraph(
+                [marginalized_value] + other_values, [joint_logp], inline=True
+            )
+            joint_logps = [
+                joint_logp_op(rv_domain_tensor[i], *other_values) for i in range(len(rv_domain))
+            ]
+
+            rv_loglike_fn = None
+            if include_samples:
+                sample_rv_outs = pm.Categorical.dist(logit_p=joint_logps)
```

Comment:
Are these `joint_logps` normalized? `pm.Categorical` won't do it under the hood

Reply:
It will when you pass `logit_p`, but the `logit_p` added to the inferencedata will still not be normalized. I think it may be more intuitive if they are but not sure.